### ISSUE TITLE
Minor issue?  I keep seeing 'node-red-project@0.0.1' when doing installs or updates to npm

### ISSUE THREAD

**Issue by dachshund-digital** [node-red/node-red#2671]
Minor issue?  I keep seeing 'node-red-project@0.0.1' when doing installs or updates to npm.  The package.json has not been updated?

 "name": "node-red-project",
  "description": "A Node-RED Project",
  "version": "0.0.1",

**Comment by hardillb** [comment]
As pointed out on the last issue you raised, you need to read and follow the instructions in the template you were presented with when you opened this issue.

**Comment by knolleary** [comment]
@dachshund-digital please use the issue template when raising issues. If you are not sure about something, please use the forum or slack to ask, rather than raise an issue (as described in the template).

In this instance, you don't describe what behaviour you expect to see. Node-RED never updates the version number in your project's package.json file - that's for you to do as and when you want to update the version number.

**Comment by dachshund-digital** [comment]
Really?  When the issue is clearly easy to find and see?  The package.json for Node Red has not been updated.  I provided the proof of that fact.  You want me to fill out a 2 or 3 page form, for this?  If you just open your own package.json for Node Red, you will see what the issue is.

**Comment by knolleary** [comment]
Part of the template says "what do you expect to happen". I'm not clear what the actual issue is here. What aspect of the package file has not been updated? The version number or subverting else?

**Comment by kuema** [comment]
@dachshund-digital: you are referring to the package.json of your Node-RED workspace/project folder, usually residing in `[userhome]/.node-red`.

The workspace folder is just like any other NodeJS package, along with metadata like name, version, etc., and keeping track of its dependencies. So you usually have a list of contrib nodes with a package version listed under `dependencies` of your package.json. Apart from that, Node-RED itself should **not** be listed there, since its your workspace folder. It should usually be installed in a separate location (either globally or a local folder).

When you run `npm install` or `npm update` in your workspace, you are installing/updating contrib nodes, not Node-RED.

The version field is never touched by NPM in any case, it's your responsibility. However, that field doesn't matter in case of a Node-RED workspace, because that package will most likely never get pushed to a registry anyway. So you could just ignore that version and keep it at 0.0.1.

There is no bug here, this is all normal (NPM) behaviour.

**Comment by dachshund-digital** [comment]
Not a bug, bad information.  How is this EVER my responsibility?  I did not create the package.json, your installation script did.  The actual package.json in the node-red hidden directory the third line of the file.  This implies that npm init was run and someone somewhere for the actual Node Red project did not update the version information... if that is not the case, then it is very misleading at a minimum.

{
  "name": "node-red-project",
  "description": "A Node-RED Project",
  "version": "0.0.1",

This information must be coming from Node Red development, no?  If I install Node Red right now.  The package.json... default will show the version of Node Red as 0.0.1.  How can this be acceptable under any context?  Node Red is already 1.1.2, or 1.1.3 rather just released right?

My point for stated the above, is I don't care who, what or why, but that clearly someone that owns core development of Node Red, should, address this issue.  Because what is happening now, for whatever reason, is not accurate, IMHO.

**Comment by knolleary** [comment]
So, you're talking about the `package.json` file in `~/.node-red` ? Nothing to do with the projects feature.

This is why we ask you to fill in the template - so that enough information is provided at the start and we don't have to ask lots of questions to understand what the issue is about.

That file is automatically created when Node-RED first starts up. Whenever nodes are installed using `npm`, `npm` expects there to be a `package.json` file present for it to update with the list of modules that are being installed. It represents the list of extra modules you have installed in *your* Node-RED instance.

To be valid, the package.json file must have a version number. We default to `0.0.1` because we have to put *something* in that field. The version number is not related to the version of Node-RED running.

Some users like to version control their userData - flows files and package.json etc. They can set the version number in the package.json file if they want to reflect points in time and versions of their flows.

It is not for the core of Node-RED to second guess what the user wants to do with the version number of their own application.


> How can this be acceptable under any context?

What exactly is not acceptable about it? We have to put a version number when the file is created. If you updated the

**Comment by dachshund-digital** [comment]
Again... to be clear, this NOT something I created.  This is from the Node Red installation its-self.  Of course if I was creating a node project... I would OWN the versioning.  But what we are talking about here is that the actual Node Red project its-self after installation has a package.json which exhibits wrong/bad information.

**Comment by knolleary** [comment]
As you have added that comment and edited it whilst I was writing my previous reply, I don't know how much of it is in response to what I just wrote.

> This is from the Node Red installation its-self.

No. It is not from the *install* of Node-RED. The package.json file for the `node-red` module is perfectly correct and is nothing to do with what you are looking at - https://github.com/node-red/node-red/blob/master/packages/node_modules/node-red/package.json

Let me try to explain again.

When you first run Node-RED it initialises the user's data directory - `~/.node-red` - with a bunch of files *for the user*. This includes:
 - the default settings file that can be customised
 - the flow and credentials files
 - a `package.json` file that will be updated by npm when you install additional node modules into *your* project

We do not expect a user to publish this user directory back to the public npm repository - you'll note we see `private: true` to reflect this fact.

> It does have meaning... it is tied to your releases, 

**No** it is not tied to our releases. It is the package.json for **your** project that is being created with Node-RED. The version of Node-RED itself is not relevant to the version of *your* project.


> Sorry but this implies that you don't follow the standard practice expected of all node developers? Surely, you can see my confusion on this point. And you are. IMHO, in error. 
>  if you can't see this as your responsibility, then you make my point... you don't follow the standards every node developer is expected to follow. 

Please pause for a minute and consider that maybe we *do* know what we are doing here and that you have misunderstood something rather than start insulting us.

**Comment by dceejay** [comment]
The actual Node-RED package.json for the application is part of the install and (on LInux) ends up at usr/lib/node_modules/node-red/package.json - where indeed it maintains it's own version number, dependencies, etc.
The one in the/your user directory (again typically) .node-red - is for your running instance of it and just contains any extra modules you have added/installed on top - as the running user may/will not have access permission to write to the global package.

**Comment by dachshund-digital** [comment]
Clearly you can't see, or maybe agree, with my perspective... but you are presenting wrong information, by not addressing this, by proxy.  Every single person that looks at the package.json file is saying... What the heck is this 0.0.1 stuff.  I can't seem to impress upon those that have replied to this thread... that it is an important issue, that you should own, address, etc.  

This is a case IMHO you should do what is right for users of Node Red.  Because this is not technically your responsibility, I get that.  But that is not the point I have, I guess failed to make, that it should be your responsibility.  It is confusing, it is misleading.  I have been asked more than once by others and peers, why this version is 0.0.1 and not matching the true release version of Node Red.  In a corporate world, with auditors, QA and QC, and more, evaluating solutions, this is a clear red flag.  So this type of thing will impact the future adoption of Node Red.  To enterprise users, it looks wrong, it is perceived as wrong.  Hence why I raised the question, issue.  If I owned Node Red, was a significant contributor, I would be asking why this happened, happens., has persisted.

And, yes, before you reply that Node Red is not targeted to serious users, but we all know that is not true.

**Comment by dceejay** [comment]
"Every single person that looks at the package.json file is saying..."

so far the total that I am aware of is 1.... is there some discussion thread somewhere else that we are missing out on ? I try to read all posts to the forum but it has been rather busy recently. 

I'm not sure how we are supposed to mange the versions of your application for you - we don't know what is a minor bug fix vs breaking change etc so how can we know which semver number to bump ? If I use VScode or Atom to edit another node.js project it doesn't touch the package.json for me - nor does it tell me which version of VScode or Atom I'm using. All we are doing is providing an initial scaffold like npm init would do for you.

Yes - we are well aware that Node-RED is used by lots of very serious users - see https://nodered.org/about/community/survey/2019/#you-and-node-red - and again this is the first time this has been mentioned.

**Comment by knolleary** [comment]
> Every single person that looks at the package.json file is saying... What the heck is this 0.0.1 stuff.

Is the explanation of: "it's the version of *your* Node-RED project. If you want to give it a different version number, then go right ahead" not enough?

> I have been asked more than once by others and peers, why this version is 0.0.1 and not matching the true release version of Node Red

... because its the version of *your* application and you have chosen not to update the version to reflect *your* application's maturity.

If you distribute a Node.js application, do you get asked why its package.json version number doesn't match the node.js version it runs on? That is the analogy to what you are saying.

Node-RED is a development tool for building applications. The versioning of the application is up to the developer using Node-RED - it is not for Node-RED to tell the user what version their own application is at.

> In a corporate world, with auditors, QA and QC, and more, evaluating solutions, this is a clear red flag. S

I agree. And if I were distributing an application I had built with Node-RED I would give it a version number that reflected its maturity and update that version whenever I distributed updated versions - just as I would when publishing updated modules to npm.

Okay, let's take a step back.

I hope I have explained that the `package.json` that gets created is for the user's application they are developing in Node-RED. Do you agree with that statement?

The question is all about what version number we put in that package.json file when we generate it.

The options would seem to be:

1. `0.0.1` as we do today
2. `1.0.0` which would match the default behavior of `npm init` if you accept its default values
3. The version of Node-RED that generated the file.
4. some other value.

Let's examine those options, remember we are just talking about the initial value put into that file that the user is free to update however they see fit.

1. `0.0.1` vs `1.0.0`. My take on this all along has been that when we first initialise the project there are no flows. It doesn't do anything at all. To give that a `1.0.0` would be, in my opinion, misleading. This is why we put in `0.0.1`. I'm happy to discuss that further. I could certainly be swayed that we should put `1.0.0` into the file by default to match the `npm init` behaviour.

2. **The version of Node-RED** - I completely disagree with this. When you run `npm init`, it doesn't put the `npm` version or `node.js` version into the package.json version. That would be completely inappropriate. The version number of the user's project is independent of the version of the runtime (aka Node-RED) .

3. Some other value... any suggestions?

**Comment by concept10** [comment]
@dachshund-digital Please go easy.  As in the corporate world, contribution rules ensure easy processes are in place to resolve errors.  Some errors are truly not the responsibility of the project you think it may be associated with.

**Comment by dachshund-digital** [comment]
@knolleary 
Appreciate your analysis.  I agree this is a sticky issue.  This is why I raised it.  Someone has to own it at some level, even if it is not strictly accurate in reference to responsibility... due to the nature of the environment, design, etc.  I think the solution that would avoid the greatest confusion is option 3.


The key problem is... it is misunderstood, by those that have no knowledge of npm concepts or structure... my management included, the QA and QC team, the internal auditors.  I have tried to explain this.

@concept10 
Actually, you point is true, but for one key point... that is not how auditors think, nor how my internal QC and QA teams think.  Let me explain.  About 10 years ago, I was dealing with IBM, they had a DNS name resolution bug, where their hardware monitoring agent would fail to report cpu, memory, etc state/status if the target monitoring server was in the local configuration file in CAPS.  If the name of the monitoring server name was in lowercase, everything worked.  The actual bug was not a bug, the best practice was and is DNS all in lower case.  But... we found the issue, and communicated it to IBM.  IBM said it is an issue in the open-source aspect of our commercial solution, i.e. monitoring agent.  IBM refused to change the open source.  My company said, no, we pay for your solution as a commercial product, regardless of the open-source elements.  Well, the auditors documented it as a finding, thus a global fortune 10 (not 50) 10 corporation total IBM fix the issue or we are pulling our 100 million hardware commitment.  What did IBM do?  They changed the open-source code, their commercial solution leveraged.  My point for this story is... you can't ignore anything, in a corporate audit/QC/QA context no matter how small, no matter who really owns it.

**Comment by knolleary** [comment]
> The perception is, the package.json file is owned and created by Node Red, because after a fresh install, it exists, that it is created as part of the installation process, is the perception. And, this is exactly what the QA/QC team review believed and saw.

How can we dispel this perception? Because it is wrong. And you are the first person in 7 years to raise it as a concern.

The file is **not** created as part of the installation process. It is created when you first run Node-RED if it doesn't already exist. That is an important distinction.

It is created in what all of the documentation calls the 'user directory' and contains all of the user's files.

If you were deploying a node-red application into production you'd have the flow files and package.json under full version control. You'd clone those files to your server then run Node-RED. Node-RED would not create a new package.json file, it'd use the one from your project.

> Why can't the Node Red installation script change the file? Why is that considered unacceptable, when it will avoid confusion? Your points/concerns noted.

To repeat again:

The `package.json` file belongs to the application the user is developing. It is the version of the **user's application**. It is up to the user to decide when they want to give their application a new version number based on the changes they make in their application.

You said it would avoid confusion if Node-RED changed the version. I really don't understand how you can say that.

If I have developed a Node-RED application and I've given it a version of `2.3.4`, I then upgrade Node-RED and run it, and suddenly my project's version number has been changed back to `1.1.3` (the latest NR release) - how is that in any way acceptable?

Answer me this: when you run `npm init` to start a new node.js project, do you expect npm to insert its own version number into your new project's package.json file? As @dceejay said, does VSCode or Atom impose their version number into your code? 

These are the exact parallels to what you are suggesting here.

I can see a place to expose an option in the Node-RED editor to let the user modify the version number via the editor and not have to edit the file manually. But that is still giving the user the responsibility of doing it.

**Comment by dachshund-digital** [comment]
Maybe because I am the first person... to actually care about Node Red as to its potential adoption in a highly qualified and audit controlled environment?  I answer to a very specific set of criteria, specific to a unique business environment.  Is it a bit crazy, sure.  Dealing with auditors is always... a bit asinine.

I can only answer your question, this way... when someone reads the package.json file, After Node Red installation, does 0.0.1 make sense?  At face value, it does not make sense.  So maybe you tackle this in light of that observation.  A small bit of code in the installation script could tweak the package.json, and the bad perception never results.

Only raised this issue because the question was asked, and when I looked at the package,json after install of Node Red, even I said to myself, that 0.0.1 is just not right.  It was only then I thought of how npm works.  Auditors deal with absolutes, not qualifications.

How about this, document the issue, and call it done.  As long as I can point to an 'official' explanation of why the package.json has 0.0.1 but the Node Red version is, say 1.1.3 for example.  That may be sufficient, at least for the internal auditors, it should be.  For the external auditing, maybe acceptable maybe not, all depends on who it is.

**Comment by HLeithner** [comment]
I am really impressed of the patience of @knolleary and the resistance of understanding of @dachshund-digital.

It's really simple Node-Red is an IDE, if you create a project in an IDE you normally give you project a version number if you don't it's likely that your IDE does this for you and starts by 1.0.0 or 0.0.1 (doesn't matter) but if will never start at the version number of the IDE and will never upgrade the version number of your project because you upgrade your IDE.

Node-Red creates your project when you install node-red which is basically a shortcut for you.

The only thing which maybe is missing to change the version number from the Webinterface (maybe in the project settings and/or in at the history commit interface).

**Comment by dceejay** [comment]
Personally I have always found most auditors to be perfectly reasonable and willing to be educated as it helps their deeper understanding. Indeed they do often have BS detectors set to high sensitivity but that is to be expected, so we stick to the facts (both good and bad) and usually have no problem. Indeed we have applications running at organisations that take project governance, security and audit **very** seriously.

The fact that the default name is "node-red-project" and the description is "A Node-RED Project" should, along with the fact that they are in your user directory space, should be adequate to explain it - though yes as @HLeithner suggests allowing editing the version in the projects UI would be a sensible enhancement - as long as you use the projects feature which I suppose all highly qualified environments would be.

**Comment by Nodi-Rubrum** [comment]
@dceejay 
My experience with auditors has been completely different, in fact, as the pressure to secure environments, deem critical to the financial and security of our enterprises, world-wide, they are under ever increasing pressure, and this is routinely impacted their flexibility to issues and risk acceptance.  One point... Node Red directory is not my directory, per the perception communicated to me, it is 'your' directory, because it is 'your' installation.  This is key to the confusion the 0.0.1 value did create from my world.

@HLeithner 
It has never been a question of resistance of understanding, it was an honest attempt to correct an issue that was raised to me, and in turn I raised it here.  I future qualified that I have different or high level of expectation from my peers and internal auditors.  This entire discussion was an attempt to eliminate an issue.  For various reasons, what could have been a simple fix from my perspective, the perspective of my world, is not such in fact.  So @knolleary itemized the options, and although I disagree with the result in principle, I still suggested what I thought, my world, could accept.  Moreover, because I am the only one that raised the issue does not mean others have not, and just not chosen to try to resolve the issue.  If others have accepted it, maybe they are not doing the same deep analysis that my world routinely does.

**Comment by dceejay** [comment]
Ok - so which platform are you installing on and how are you installing ?
If using the local install - https://nodered.org/docs/getting-started/local then at the point the sudo npm install has finished there is no user directory nor package.json. So they are not "ours". As soon as you run Node-RED they do get created and initialised. This is same behaviour as many other apps. At this point they are "yours" as they are individual to the user running Node-RED. Yes I'm being pedantic - as auditors are :-)

If you use linux and the install script then yes - it does go further. In order to customise it for Pi etc we do install some extra nodes, and install some scripts to make stopping and starting easier, and as part of this extra work we do initialise the user directory as that is where we store the extra nodes for you. 

You are of course free to not use the script (if you do) and can thus show that the install doesn't create anything in "your" space. 

Alternatively you can copy and edit the script to initialise the package.json anyhow you wish - it is open source after all - and indeed you could then make an installer that met all your corporate needs, and any other local customisations you may wish. I know my employer certainly has their own application repository of approved apps with custom configurations etc.

**Comment by Nodi-Rubrum** [comment]
Right, most recently with PXE support on Pi 4 devices, we began looking at how we could develop Pi devices as remote monitoring devices, via various methods, which I will not itemize here.  Since Node Red has history with Pi devices, that was the start of serious evaluation from our end.

**Comment by dceejay** [comment]
Ok - so yes - I assume you use the script so yes feel free to customise it - especially around line 355

**Comment by Nodi-Rubrum** [comment]
Noted... Do you still plan to note anything in the documentation officially, so we have a point-to-reference?  It always is easier with the audit reviews to reference official sources when possible.  I can't use this thread unfortunately because it has not official standing that would be recognized at our end.

**Comment by dceejay** [comment]
maybe here ? https://nodered.org/docs/user-guide/runtime/adding-nodes

**Comment by knolleary** [comment]
We can add some docs that better describe the user directory and what it contains. That would be useful in the context of documenting how to properly backup your data if you aren't using the projects feature.

We could also change the description field of the package.json we generate along the lines of what @dceejay has suggested to indicate it was generated for the user, and include the version of nr that did it.

**Comment by dachshund-digital** [comment]
Thank you.
